lib/autobuild/package.rb

```diff
@@ -584,6 +584,15 @@ def post_install(*args, &block)
             end
         end
 
+        def fingerprint(recursive: true)
+            concatenate_fingerprints = importer.fingerprint(self)
+            dependencies.each do |pkg_name|
```

Comment:
A more ruby-ish version is with `inject`

~~~
concatenate_fingerprints = 
    dependencies.inject(importer.fingerprint(self)) do |concat, pkg_name|
         pkg = Autobuild::Package[pkg_name]
         concat + pkg.importer.fingerprint(pkg)
    end
end
~~~

Note that this implementation is sensitive to the order in `dependencies`. Sort before you fingerprint.

And you haven't implemented the recursive version. You could either call recursively using pkg.fingerprint instead of pkg.importer.fingerprint, or use `all_dependencies`, which returns the list of dependencies recursively. I would personally prefer the former, because it would allow you to pass a `memo` hash to memoize the fingerprint values, to not repeatedly calculate the same fingerprints over and over again when fingerprinting a lot of packages at the same time.

Reply:
What is this for? Isn't doing sha1 validation during import phase enough?

Reply:
They idea is to have a unique fingerprint of a package including it's dependencies, so the method hashes them all together.

Reply:
I see that but what's the goal (besides integrity check)? Moreover, why do that recursively if you are already doing it per package?

Reply:
We're caching build artifacts within the CI, so that one can rebuild only the parts that have been modified. Rather than having to re-do the recursive computation, we provide it, which allows to easily answer the question "has something changed with this package ?" externally without having to walk the dependency tree.

Basically, the logic is:
- read fingerprint from build manifest
- check whether there is a `$FINGERPRINT.tar.xz` in the cache